clients/include/rocblas_arguments.hpp

```diff
@@ -324,4 +328,216 @@ static_assert(std::is_trivial<Arguments>{},
               "Arguments is not a trivial type, and thus is "
               "incompatible with C.");
 
+enum arg_type
+{
+    e_M,
+    e_N,
+    e_K,
+    e_lda,
+    e_ldb,
+    e_ldc,
+    e_ldd,
+    e_a_type,
+    e_b_type,
+    e_c_type,
+    e_d_type,
+    e_compute_type,
+    e_incx,
+    e_incy,
+    e_incb,
+    e_incd,
+    e_alpha,
+    e_beta,
+    e_transA,
+    e_transB,
+    e_side,
+    e_uplo,
+    e_diag,
+    e_batch_count,
+    e_stride_a,
+    e_stride_b,
+    e_stride_c,
+    e_stride_d,
+    e_stride_x,
+    e_stride_y,
+    e_algo,
+    e_solution_index,
+
+};
+
+#define CTOKEN(n) const char* const c_tok_##n = #n
+CTOKEN(M);
+CTOKEN(N);
+CTOKEN(K);
+CTOKEN(lda);
+CTOKEN(ldb);
+CTOKEN(ldc);
+CTOKEN(ldd);
+CTOKEN(a_type);
+CTOKEN(b_type);
+CTOKEN(c_type);
+CTOKEN(d_type);
+CTOKEN(compute_type);
+CTOKEN(incx);
+CTOKEN(incy);
+CTOKEN(incb);
+CTOKEN(incd);
+CTOKEN(alpha);
+CTOKEN(beta);
+CTOKEN(transA);
+CTOKEN(transB);
+CTOKEN(side);
+CTOKEN(uplo);
+CTOKEN(diag);
+CTOKEN(batch_count);
+CTOKEN(stride_a);
+CTOKEN(stride_b);
+CTOKEN(stride_c);
+CTOKEN(stride_d);
+CTOKEN(stride_x);
+CTOKEN(stride_y);
+CTOKEN(algo);
+CTOKEN(solution_index);
+#undef CTOKEN
+
+class ArgumentModel
+{
+public:
+    ArgumentModel(const std::vector<arg_type>& params);
+    virtual ~ArgumentModel() {}
+
+    bool hasParam(arg_type a);
+
+    template <typename T>
+    static void print_value(std::ostream& str, const T& x)
+    {
+        Arguments::print_value(str, x);
+    }
+
+    // Float16 output
+    static void print_value(std::ostream& str, rocblas_half x)
+    {
+        Arguments::print_value(str, double(x));
+    }
+
+    template <typename T>
+    void log_args(std::ostream&    str,
+                  const Arguments& args,
+                  double           gpu_us,
+                  double           gflops,
+                  double           gpu_bytes = 0,
+                  double           cpu_us    = 0,
+                  double           norm1     = 0,
+                  double           norm2     = 0);
+
+    virtual void log_perf(std::stringstream& name_str,
+                          std::stringstream& val_str,
+                          const Arguments&   arg,
+                          double             gpu_us,
+                          double             gflops,
+                          double             gpu_bytes,
+                          double             cpu_us,
+                          double             norm1,
+                          double             norm2);
+
+protected:
+    std::vector<arg_type> m_args;
```

Comment:
This adds a lot of extra overhead, when it would be simpler to use composition/inheritance and CTRP with `Arguments` instead of a dynamically allocated `vector`.

Reply:
If you want to print out a specific argument set, it's best to list name, value pairs in an arbitrary-length print function call, making adjustments to names and values (such as alpha, beta) as necessary.

Reply:
A `std::tuple`, listing names and values/references, is just a container representation of it.

Reply:
This is for test setup, so a single static instance can be created per function so almost zero relative overhead compared to other things.

Reply:
> This is for test setup, so a single static instance can be created per function so almost zero relative overhead compared to other things.

Then this should be in another tool, such as `rocblas_gentest.py`.